browser-polyfill.js

```diff
@@ -0,0 +1,340 @@
+     *
+     * @param {object} [metadata = {}]
+     *        An object tree containing metadata used to automatically generate
+     *        Promise-based wrapper functions for asynchronous. Any function in
+     *        the `target` object tree which has a corresponding metadata object
+     *        in the same location in the `metadata` tree is replaced with an
+     *        automatically-generated wrapper function, as described in
+     *        {@see wrapAsyncFunction}
+     *
+     * @returns {Proxy<object>}
+     */
+    const wrapObject = (target, wrappers = {}, metadata = {}) => {
+      let cache = Object.create(null);
+
+      let handlers = {
+        has(target, prop) {
+          return prop in target || prop in cache;
+        },
+
+        get(target, prop, receiver) {
+          if (prop in cache) {
+            return cache[prop];
+          }
+
+          if (!(prop in target)) {
+            return undefined;
+          }
+
+          let value = target[prop];
+
+          if (typeof value == "function") {
+            // This is a method on the underlying object. Check if we need to do
+            // any wrapping.
+
+            if (typeof wrappers[prop] === "function") {
+              // We have a special-case wrapper for this method.
+              value = wrapMethod(target, target[prop], wrappers[prop]);
+            } else if (hasOwnProperty(metadata, prop)) {
+              // This is an async method that we have metadata for. Create a
+              // Promise wrapper for it.
+              let wrapper = wrapAsyncFunction(prop, metadata[prop]);
+              value = wrapMethod(target, target[prop], wrapper);
+            } else {
+              // This is a method that we don't know or care about. Return the
+              // original method, bound to the underlying object.
+              value = value.bind(target);
+            }
+          } else if (typeof value == "object" &&
+                     (hasOwnProperty(wrappers, prop) ||
+                      hasOwnProperty(metadata, prop))) {
+            // This is an object that we need to do some wrapping for the children
+            // of. Create a sub-object wrapper for it with the appropriate child
+            // metadata.
+            value = wrapObject(value, wrappers[prop], metadata[prop]);
+          } else {
+            // We don't need to do any wrapping for this property,
+            // so just forward all access to the underlying object.
+            Object.defineProperty(cache, prop, {
+              configurable: true,
+              enumerable: true,
+              get() {
+                return target[prop];
+              },
+              set(value) {
+                target[prop] = value;
+              },
+            });
+
+            return value;
+          }
+
+          cache[prop] = value;
+          return value;
+        },
+
+        set(target, prop, value, receiver) {
```

Comment:
`receiver` is currently unused, how about omitting it from the listed parameters? (or adding an inline comment for it to explain that it is ignored on purpose because we don't need to cover that particular scenario).